### DISCUSSION TITLE
Failing the installation tests on Mac

### DISCUSSION THREAD

**Discussion by hwoo749** [firedrakeproject/firedrake#2885] [Firedrake support]
Hi there,

I'm using Homebrew Python 3.11.3 on macOS 12.15.1.

After installing firedrake and running the test installation command: pytest tests/regression/ -k "poisson_strong or stokes_mini or dg_advection", I get the following errors.

'''
(firedrake) hw@wifi-452-91-423-678 firedrake % pytest tests/regression/ -k "poisson_strong or stokes_mini or dg_advection"

========================================================== test session starts ==========================================================
platform darwin -- Python 3.11.3, pytest-7.3.1, pluggy-1.0.0
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/hw/firedrake/src/firedrake
configfile: setup.cfg
plugins: anyio-3.6.2, benchmark-4.0.0
collected 5 items / 154 errors                                                                                                          
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/_pytest/main.py", line 269, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>                          ^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/_pytest/main.py", line 322, in _main
INTERNALERROR>     config.hook.pytest_collection(session=session)
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/pluggy/_hooks.py", line 265, in __call__
INTERNALERROR>     return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/pluggy/_manager.py", line 80, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/pluggy/_callers.py", line 60, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/pluggy/_result.py", line 60, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/pluggy/_callers.py", line 39, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/_pytest/main.py", line 333, in pytest_collection
INTERNALERROR>     session.perform_collect()
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/_pytest/main.py", line 668, in perform_collect
INTERNALERROR>     hook.pytest_collection_modifyitems(
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/pluggy/_hooks.py", line 265, in __call__
INTERNALERROR>     return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/pluggy/_manager.py", line 80, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/pluggy/_callers.py", line 60, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/pluggy/_result.py", line 60, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/Users/hw/firedrake/lib/python3.11/site-packages/pluggy/_callers.py", line 39, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/hw/firedrake/src/firedrake/tests/conftest.py", line 20, in pytest_collection_modifyitems
INTERNALERROR>     from firedrake.utils import complex_mode, SLATE_SUPPORTS_COMPLEX
INTERNALERROR> ModuleNotFoundError: No module named 'firedrake'

''' 

After trying this command: python -m pytest tests/regression/ -k "poisson_strong or stokes_mini or dg_advection"

The final error message says module not found for petsc4py rather than firedrake : INTERNALERROR>     import petsc4py
INTERNALERROR> ModuleNotFoundError: No module named 'petsc4py'

I have installed PETSc and uninstalled anaconda - how do I go about passing the tests and fixing this issue? 

Thanks in advance.

**Comment by dham** [comment]
We should first check that your Firedrake install succeeded. Please post `firedrake-install.log`, which you will find next to the top firedrake folder on your computer. I.e. `/Users/hw/firedrake-install.log`

**Reply by hwoo749** [reply]
Thank you! Here is the firedrake-install.log

[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/11348313/firedrake-install.log)

**Reply by dham** [reply]
Your Firedrake install didn't complete. It also doesn't look like it crashed, but rather that it was interrupted during the PETSc install. 

You should dectivate the venv, remove the top Firedrake folder and try again. If it succeeds you will see a message at the end saying "Successfully installed Firedrake." followed by instructions for activating the venv and upgrading Firedrake. If it doesn't succeed then please post the resulting log again. Note that depending on your system, building Petsc is likely to take tens of minutes.

**Reply by hwoo749** [reply]
Thank you for the quick response. I think PETSc was previously installed on my Mac, but it could have been interrupted then, as it took several hours and my WiFi connection dropped at times. Would you suggest uninstalling PETSc and solely installing it through firedrake? Or how can I check that PETSc was installed correctly? 

I followed your instructions but did not get the "Successfully installed Firedrake." message. This was my output after deactivating the venv and removing firedrake, then starting again:

'''
harrietwoodward@Harriets-MacBook-Air ~ % curl -O https://raw.githubusercontent.com/firedrakeproject/firedrake/master/scripts/firedrake-install
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 93388  100 93388    0     0   178k      0 --:--:-- --:--:-- --:--:--  181k
harrietwoodward@Harriets-MacBook-Air ~ % python3 firedrake-install
Running firedrake-install
Installing command line tools...
Checking brew doctor...
Installing required packages via homebrew...
Required package 'gcc' is already installed via Homebrew.
Required package 'autoconf' is already installed via Homebrew.
Required package 'make' is already installed via Homebrew.
Required package 'automake' is already installed via Homebrew.
Required package 'cmake' is already installed via Homebrew.
Required package 'libtool' is already installed via Homebrew.
Required package 'boost' is already installed via Homebrew.
Required package 'openblas' is already installed via Homebrew.
Creating firedrake venv in '/Users/harrietwoodward/firedrake'.
BLAS config is {'BLAS': '/usr/local/opt/openblas', 'OPENBLAS': '/usr/local/opt/openblas', 'LDFLAGS': '-L/usr/local/opt/gcc/lib/gcc/12'}
Configuration saved to /Users/harrietwoodward/firedrake/.configuration.json
Cloning firedrake

Failed to clone firedrake using ssh, falling back to https.
Successfully cloned repository firedrake.
Checking out branch master
Successfully checked out branch master
Updating submodules.
Successfully updated submodules.
Creating firedrake-update script.
We use pybind11==2.9.2 until the issue with pybind11==2.10.0 is resolved.

Cloning the dependencies of firedrake
Cloning ufl

Failed to clone ufl using ssh, falling back to https.
Successfully cloned repository ufl.
Checking out branch master
Successfully checked out branch master
Updating submodules.
Successfully updated submodules.
Cloning fiat

Failed to clone fiat using ssh, falling back to https.
Successfully cloned repository fiat.
Checking out branch master
Successfully checked out branch master
Updating submodules.
Successfully updated submodules.
Cloning FInAT

Failed to clone FInAT using ssh, falling back to https.
Successfully cloned repository FInAT.
Checking out branch master
Successfully checked out branch master
Updating submodules.
Successfully updated submodules.
Cloning tsfc

Failed to clone tsfc using ssh, falling back to https.
Successfully cloned repository tsfc.
Checking out branch master
Successfully checked out branch master
Updating submodules.
Successfully updated submodules.
Cloning PyOP2

Failed to clone PyOP2 using ssh, falling back to https.
Successfully cloned repository PyOP2.
Checking out branch master
Successfully checked out branch master
Updating submodules.
Successfully updated submodules.
Cloning pyadjoint

Failed to clone pyadjoint using ssh, falling back to https.
Successfully cloned repository pyadjoint.
Checking out branch master
Successfully checked out branch master
Updating submodules.
Successfully updated submodules.
Cloning petsc

Failed to clone petsc using ssh, falling back to https.
Successfully cloned repository petsc.
Checking out branch firedrake
Successfully checked out branch firedrake
Updating submodules.
Successfully updated submodules.
Cloning pytest-mpi

Failed to clone pytest-mpi using ssh, falling back to https.
Successfully cloned repository pytest-mpi.
Checking out branch main
Successfully checked out branch main
Updating submodules.
Successfully updated submodules.
Cloning the dependencies of PyOP2
Cloning COFFEE

Failed to clone COFFEE using ssh, falling back to https.
Successfully cloned repository COFFEE.
Checking out branch master
Successfully checked out branch master
Updating submodules.
Successfully updated submodules.
Cloning loopy

Failed to clone loopy using ssh, falling back to https.
Successfully cloned repository loopy.
Checking out branch main
Successfully checked out branch main
Updating submodules.
Successfully updated submodules.
Installing petsc/
Downloading Eigen from 'https://github.com/eigenteam/eigen-git-mirror/archive/3.3.3.tar.gz' to 'eigen-3.3.3.tgz'
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 1348, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/client.py", line 1282, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/client.py", line 1328, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/client.py", line 1277, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/client.py", line 1037, in _send_output
    self.send(msg)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/client.py", line 975, in send
    self.connect()
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/client.py", line 1454, in connect
    self.sock = self._context.wrap_socket(self.sock,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ssl.py", line 513, in wrap_socket
    return self.sslsocket_class._create(
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ssl.py", line 1071, in _create
    self.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ssl.py", line 1342, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:997)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/harrietwoodward/firedrake-install", line 1885, in <module>
    install("petsc/")
  File "/Users/harrietwoodward/firedrake-install", line 1030, in install
    build_and_install_petsc()
  File "/Users/harrietwoodward/firedrake-install", line 1142, in build_and_install_petsc
    urllib.request.urlretrieve(url, filename=tarball)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 241, in urlretrieve
    with contextlib.closing(urlopen(url, data)) as fp:
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 519, in open
    response = self._open(req, data)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 536, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 1391, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 1351, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:997)>


Install log saved in /Users/harrietwoodward/firedrake-install.log

'''

I am a bit of a newcomer to this, so apologies if I am misunderstanding. Below is the new install log.

[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/11358589/firedrake-install.log)

**Reply by dham** [reply]
Firedrake will build its own PETSc unless you tell it not to (and you probably don't want to do that). It won't be affected by any other PETSc you have installed.

This time your Firedrake install failed because you used the Python that ships with your Mac rather than installing python using homebrew. This can be seen from paths like this one: `/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py` which start with `/Library/Frameworks/Python.framework` rather than something like `/opt/homebrew/bin/python3`. This fails because the Mac Python has a broken ssl and hence can't download files over the network. You should ensure you have homebrew Python installed and are using that.

**Reply by hwoo749** [reply]
Hi there, I removed and then reinstalled Python through homebrew, then tried the firedrake install again. This time the PETSc install took a day to install which seemed promising, but I still ended up with the same errors. I will post the errors and install log. 

I got these errors at the end of the PETSc install: 

'''
Installing pip dependencies for firedrake
Traceback (most recent call last):
  File "/Users/harrietwoodward/firedrake-install", line 1898, in <module>
    pip_requirements(p, compiler_env)
  File "/Users/harrietwoodward/firedrake-install", line 1001, in pip_requirements
    run_pip_install_wrap(reqs, parallel_compiler_env)
  File "/Users/harrietwoodward/firedrake-install", line 994, in run_pip_install_wrap
    run_pip_install([req])
  File "/Users/harrietwoodward/firedrake-install", line 975, in run_pip_install
    check_call(pipinstall + pipargs)
  File "/Users/harrietwoodward/firedrake-install", line 663, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/Users/harrietwoodward/firedrake/bin/python', '-m', 'pip', 'install', '--no-build-isolation', '--no-binary', 'mpi4py,randomgen,islpy,numpy', '-vvv', 'pytest-xdist']' returned non-zero exit status 2.

'''

Then before the testing I got this message which I haven't gotten before:
'''
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
'''

Here is the install log: 

[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/11378121/firedrake-install.log)

**Reply by connorjward** [reply]
The log is showing errors like:
```
pip._vendor.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out.
```
and
```
TimeoutError: The read operation timed out
```

It seems like you were installing with a flaky internet connection. Trying again may do the trick.

I also notice that `brew doctor` is raising a number of warnings. It would be good to resolve these too.

**Reply by hwoo749** [reply]
Hello, thank you for your help, I followed your instructions, and this install was definitely more successful! The only error I'm left with is this one:

========================================== ERRORS ==========================================
_____________________ ERROR collecting tests/regression/test_netgen.py _____________________
ImportError while importing test module '/Users/harrietwoodward/firedrake/src/firedrake/tests/regression/test_netgen.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/local/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/regression/test_netgen.py:2: in <module>
    from netgen.geom2d import SplineGeometry
E   ModuleNotFoundError: No module named 'netgen'
================================= short test summary info ==================================
ERROR tests/regression/test_netgen.py
!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!
============================ 3121 deselected, 1 error in 55.72s ============================
(firedrake) harrietwoodward@Harriets-MacBook-Air firedrake % 

How would I go about fixing this?

**Reply by connorjward** [reply]
Netgen is an optional dependency of Firedrake. Unless you passed `--netgen` to `firedrake-install` this is expected behaviour.
